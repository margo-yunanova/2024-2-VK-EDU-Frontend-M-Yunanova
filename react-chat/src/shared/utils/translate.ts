const checkResponse = (res: Response) =>
  res.ok ? res.json() : Promise.reject(`Ошибка: ${res.status}`);

interface ITranslatedResponse {
  responseData: {
    translatedText: string;
    match: number;
  };
}

type TTranslate = (text: string, from: string, to: string) => Promise<string>;

export const translate: TTranslate = async (text, from, to) => {
  const params = new URLSearchParams();
  params.append('q', text);
  params.append('langpair', `${from}|${to}`);

  try {
    const translatedResponse: ITranslatedResponse = await checkResponse(
      await fetch(
        `https://api.mymemory.translated.net/get?${params.toString()}`,
      ),
    );

    return translatedResponse.responseData.translatedText;
  } catch (error) {
    console.error(error);
    throw new Error('Translation failed');
  }
};
