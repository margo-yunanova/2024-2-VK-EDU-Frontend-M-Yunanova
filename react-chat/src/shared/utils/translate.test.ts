import { describe, expect, test } from 'vitest';

import { translate } from './translate';

describe('translate', () => {
  test('should return translated text', async () => {
    const response = await translate('Hello world', 'en', 'ru');
    expect(response).toBe('Привет мир!');
  });
});
