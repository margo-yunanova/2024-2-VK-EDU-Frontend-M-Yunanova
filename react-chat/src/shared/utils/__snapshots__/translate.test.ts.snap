// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`translate > should call fetch with correct search params 1`] = `
[MockFunction spy] {
  "calls": [
    [
      "https://api.mymemory.translated.net/get?q=Hello+world&langpair=en%7Cru",
    ],
    [
      "https://api.mymemory.translated.net/get?q=Hello+&langpair=en%7Cru",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
